src/mlpack/core/data/save_impl.hpp

```diff
@@ -38,220 +39,177 @@ bool Save(const std::string& filename,
-      Log::Fatal << "Cannot open file '" << filename << "' for writing. "
-          << "Save failed." << std::endl;
-    else
-      Log::Warn << "Cannot open file '" << filename << "' for writing; save "
-          << "failed." << std::endl;
-
-    return false;
-  }
-
-  // Try to save the file.
-  Log::Info << "Saving " << stringType << " to '" << filename << "'."
-      << std::endl;
-
-  // Transpose the matrix.
-  if (transpose)
-  {
-    arma::Mat<eT> tmp = trans(matrix);
-
-#ifdef ARMA_USE_HDF5
-    // We can't save with streams for HDF5.
-    const bool success = (saveType == FileType::HDF5Binary) ?
-        tmp.save(filename, ToArmaFileType(saveType)) :
-        tmp.save(stream, ToArmaFileType(saveType));
-#else
-    const bool success = tmp.save(stream, ToArmaFileType(saveType));
-#endif
-    if (!success)
-    {
-      Timer::Stop("saving_data");
-      if (fatal)
-        Log::Fatal << "Save to '" << filename << "' failed." << std::endl;
-      else
-        Log::Warn << "Save to '" << filename << "' failed." << std::endl;
-
-      return false;
-    }
-  }
-  else
-  {
-#ifdef ARMA_USE_HDF5
-    // We can't save with streams for HDF5.
-    const bool success = (saveType == FileType::HDF5Binary) ?
-        matrix.save(filename, ToArmaFileType(saveType)) :
-        matrix.save(stream, ToArmaFileType(saveType));
-#else
-    const bool success = matrix.save(stream, ToArmaFileType(saveType));
-#endif
-    if (!success)
-    {
-      Timer::Stop("saving_data");
-      if (fatal)
-        Log::Fatal << "Save to '" << filename << "' failed." << std::endl;
-      else
-        Log::Warn << "Save to '" << filename << "' failed." << std::endl;
-
-      return false;
-    }
-  }
+  MatrixOptions opts;
+  opts.Fatal() = fatal;
+  opts.NoTranspose() = !transpose;
+  opts.Format() = inputSaveType;
 
-  Timer::Stop("saving_data");
+  return Save(filename, matrix, opts);
+}
 
-  // Finally return success.
-  return true;
+template<typename MatType, typename DataOptionsType>
+bool Save(const std::string& filename,
+          const MatType& matrix,
+          const DataOptionsType& opts,
+          std::enable_if_t<IsArma<MatType>::value || IsSparseMat<MatType>::value>*)
+{
+  //! just use default copy ctor with = operator and make a copy.
+  DataOptionsType copyOpts(opts);
+  return Save(filename, matrix, copyOpts);
 }
 
-// Save a Sparse Matrix
-template<typename eT>
+/*
+ * Add this SFINAE in here because the compiler is so stupid that it is not
+ * able to distinguish between these two:
+ *
+ *  data::Save(filename, "model", *output);
+ *
+ *  and 
+ *
+ *  data::Save(filename, matrix, opts);
+ *  
+ * The second SFINAE is added because the compiler is bot able to see the
+ * difference between:
+ *
+ *  data::Save(filename, Row/Col, fatal);
+ *
+ *  and 
+ *
+ *  data::Save(filename, Row/Col, Opts);
+ *
+ * This SFINAE is temporary and must be removed after the integration of stage 3 or
+ * when the compiler becomes more intelligent.
```

Comment:
We could be waiting a long time for that second thing! :smile:

Reply:
In 30 yrs, possibly.